### PR TITLE
Components for auto-reply feature

### PR DESCRIPTION
**Pull request by vincaslt**
Auto-reply functionality for: https://github.com/ProtonMail/protonmail-settings/issues/10

**Related PRs:**

- `protonmail-settings`: https://github.com/ProtonMail/protonmail-settings/pull/29
- `design-system`: https://github.com/ProtonMail/design-system/pull/114 ✅
- `proton-shared`: https://github.com/ProtonMail/proton-shared/pull/16 ✅

Screenshot 2019-05-08 at 16 32 12

### CODE REVIEW

File: components/input/RichTextEditor.js [hunk 1]
```diff
@@ -1,5 +1,10 @@
 import React from 'react';
 import ReactQuill from 'react-quill';
+import Quill from 'quill';
+
+const Block = Quill.import('blots/block');
```
Comment (EpokK) [reply components/input/RichTextEditor.js]: Yeah, I think for now that's OK and I'm not seeing a case where `p` is useful for us.

File: containers/autoReply/AutoReplyTemplate.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                    <InfoLine label={c('Label').t`Duration`}>{durationLabel}</InfoLine>
+                    <InfoLine label={c('Label').t`Start`}>{formatTime(autoresponder.StartTime * 1000)}</InfoLine>
+                    <InfoLine label={c('Label').t`End`}>{formatTime(autoresponder.EndTime * 1000)}</InfoLine>
+                    <InfoLine label={c('Label').t`Timezone`}>{timezone}</InfoLine>
+                    <InfoLine plain label={c('Label').t`Message`}>
+                        <div dangerouslySetInnerHTML={{ __html: autoresponder.Message }} />
+                    </InfoLine>
+                </tbody>
+            </table>
+            <div className="flex w100">
```
Comment (EpokK) [inline containers/autoReply/AutoReplyTemplate.js]: Why do you need `flex w100` here?

File: components/select/TimeSelect.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    const hours = hour * HOUR;
+    return [
+        ...options,
+        { text: `${formattedHour}:00`, value: hours },
+        { text: `${formattedHour}:30`, value: hours + HOUR / 2 }
+    ];
+}, []);
+
+const TimeSelect = ({ onChange, value, ...rest }) => {
+    const handleChange = (e) => onChange(parseInt(e.target.value, 10));
```
Comment (EpokK) [inline components/select/TimeSelect.js]:
```suggestion
    const handleChange = ({ target }) => onChange(+target.value);
```

File: containers/autoReply/AutoReplyModal.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+const AutoReplyModal = ({ onClose }) => {
+    const [{ AutoResponder }] = useMailSettings();
+    const { model, updateModel, toAutoResponder } = useAutoReplyForm(AutoResponder);
+    const { request } = useApiWithoutResult(updateAutoresponder);
+    const { call } = useEventManager();
+
+    const handleSubmit = async () => {
+        await request(toAutoResponder(model));
+        call();
```
Comment (EpokK) [inline containers/autoReply/AutoReplyModal.js]: We should display a notification to say "Auto-reply updated". For notification you can use `useNotifications`.

``` js
const { createNotification } = useNotifications();
createNotification({ text: c('Success').t`Auto-reply updated` });
```

File: containers/autoReply/AutoReplyForm/useAutoReplyForm.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    DaysSelected: model.daysOfWeek,
+    Zone: model.timeZone,
+    Subject: model.subject,
+    IsEnabled: model.enabled,
+    StartTime: (model.startDate + model.startTime) / 1000,
+    EndTime: (model.endDate + model.endTime) / 1000
+});
+
+const useAutoReplyForm = (AutoResponder) => {
+    const [model, setModel] = useState(() => toModel(AutoResponder));
```
Comment (EpokK) [inline containers/autoReply/AutoReplyForm/useAutoReplyForm.js]: Why the model is a `function`? Why about an `object`?

File: components/input/RichTextEditor.js (hunk 1 above)
Comment (vincaslt) [inline components/input/RichTextEditor.js]: Would be nice to find a better way to set defaults for this component. Now this sets div tag as a default element for every instance of RichTextEditor. Might be useful to allow configuring individually.

File: containers/autoReply/AutoReplyForm/useAutoReplyForm.js (hunk 5 above)
Comment (vincaslt) [reply containers/autoReply/AutoReplyForm/useAutoReplyForm.js]: My model is an object. When a function is passed to useState - it only calls it once on initial render. If I wrote `useState(toModel(AutoResponder));` my toModel function would get called every time the component re-rendered.

https://reactjs.org/docs/hooks-reference.html#lazy-initial-state

File: containers/autoReply/AutoReplyTemplate.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            ).text;
+            return c('Duration').t`Every ${dayOfMonth} @ ${hours}`;
+        } else if (autoresponder.Repeat === AutoReplyDuration.PERMANENT) {
+            return c('Duration').t`n/a`;
+        }
+    };
+
+    return (
+        <div className="bordered-container p2 mw650p">
+            <table>
```
Comment (EpokK) [inline containers/autoReply/AutoReplyTemplate.js]: Why not using a structure with `Row`, `Label` and `Field`?

Comment (vincaslt) [reply containers/autoReply/AutoReplyTemplate.js]: From the design it seemed that the all of the labels must have the same width and take it from the widest one. That's basically a table. If I use Rows, labels would have different sizes or I would need to set a fixed size for them. I've hard opinion on this, just it seemed to be simpler this way.

**Pull request merged**